File: dsa-learning/dsa-python-code/com/dsa/leetcode/labuladong/backtrack/P78_Subsets.py
from typing import List


class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        ans = []
        path = []
        self.backtrack(nums, 0, path, ans)
        return ans

    def backtrack(self, nums, start, path, ans):
        ans.append(path[:])
        if start == len(nums):
            return
        for i in range(start, len(nums)):
            path.append(nums[i])
            self.backtrack(nums, i + 1, path, ans)
            path.pop()

File: dsa-learning/dsa-java-code/src/main/java/com/dsa/leetcode/labuladong/backtrack/P78_Subsets.java
package com.dsa.leetcode.labuladong.backtrack;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

/**
 * @since 2023/11/18 10:41
 */
public class P78_Subsets {
    List<List<Integer>> ans = new LinkedList<>();
    LinkedList<Integer> path = new LinkedList<>();

    public List<List<Integer>> subsets(int[] nums) {
        backtrack(nums, 0);
        return ans;
    }

    private void backtrack(int[] nums, int start) {
        // 前序收集
        ans.add(new LinkedList<>(path));
        // 终止条件可以不写
        if (start == nums.length) {
            return;
        }
        for (int i = start; i < nums.length; i++) {
            path.add(nums[i]);
            backtrack(nums, i + 1);
            path.removeLast();
        }
    }

    public static void main(String[] args) {
        int[] nums = {1, 2, 3};
        P78_Subsets subsets = new P78_Subsets();
        System.out.println(subsets.subsets(nums));
    }

}
